pkg/internal/storage/repo/outputs.go

```diff
@@ -58,54 +59,56 @@ func (o *Outputs) FindOutputsByTransactionID(ctx context.Context, transactionID
 		return nil, fmt.Errorf("failed to find outputs for transactionID: %d: %w", transactionID, err)
 	}
 
-	return slices.Map(outputRows, o.mapModelToTableOutput), nil
+	return slices.Map(outputRows, o.mapModelToOutputEntity), nil
 }
 
-func (o *Outputs) mapModelToTableOutput(model *models.Output) *wdk.TableOutput {
-	output := &wdk.TableOutput{
-		CreatedAt:          model.CreatedAt,
-		UpdatedAt:          model.UpdatedAt,
-		OutputID:           model.ID,
-		UserID:             model.UserID,
-		TransactionID:      model.TransactionID,
-		BasketName:         model.BasketName,
-		Spendable:          model.Spendable,
-		Change:             model.Change,
-		OutputDescription:  model.Description,
-		Vout:               model.Vout,
-		Satoshis:           model.Satoshis,
-		ProvidedBy:         model.ProvidedBy,
-		Purpose:            model.Purpose,
-		Type:               model.Type,
-		DerivationPrefix:   model.DerivationPrefix,
-		DerivationSuffix:   model.DerivationSuffix,
-		CustomInstructions: model.CustomInstructions,
-		LockingScript:      model.LockingScript,
-		SenderIdentityKey:  model.SenderIdentityKey,
-	}
-	if model.Transaction != nil {
-		output.TxID = model.Transaction.TxID
-	}
-	return output
-}
-
-func (o *Outputs) ListAndCountOutputs(ctx context.Context, filter entity.ListOutputsFilter) ([]*wdk.TableOutput, int64, error) {
+func (o *Outputs) ListAndCountOutputs(ctx context.Context, filter entity.ListOutputsFilter) ([]*entity.Output, int64, error) {
 	var outputs []*models.Output
 	var total int64
 
 	if err := o.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
 		query := tx.
 			Model(&models.Output{}).
-			Where("user_id = ?", filter.UserID)
+			Where("user_id = ?", filter.UserID).
+			Preload("Transaction", func(db *gorm.DB) *gorm.DB {
+				return db.Select("id, tx_id")
+			})
+
+		var omitFields []string
+
+		if !filter.IncludeLockingScripts {
+			omitFields = append(omitFields, "locking_script")
+		}
+
+		if !filter.IncludeCustomInstructions {
+			omitFields = append(omitFields, "custom_instructions")
+		}
+
+		if len(omitFields) > 0 {
+			query = query.Omit(omitFields...)
+		}
 
 		if filter.Basket != "" {
 			query = query.Where("basket_name = ?", filter.Basket)
 		}
 
-		if filter.IncludeTXID {
-			query = query.Preload("Transaction", func(db *gorm.DB) *gorm.DB {
-				return db.Select("id, tx_id")
-			})
+		if filter.IncludeTags {
+			query = query.Preload("Tags")
+		}
+
+		if len(filter.Tags) > 0 {
+			labelSubQuery := tx.Model(&models.OutputTags{}).
+				Select("output_id").
+				Where("tag_name IN ?", filter.Tags).
+				Where("tag_user_id = ?", filter.UserID)
+
+			if filter.TagsQueryMode == defs.QueryModeAll {
+				labelSubQuery = labelSubQuery.
+					Group("output_id").
+					Having("COUNT(DISTINCT tag_name) = ?", len(filter.Tags))
+			}
+
+			query = query.Where("id IN (?)", labelSubQuery)
 		}
```

Comment:
```suggestion
		query = o.applyTagFilter(query, filter)
```

```go
func (o *Outputs) applyTagFilter(db *gorm.DB, filter entity.ListOutputsFilter) *gorm.DB {
    if len(filter.Tags) == 0 {
        return db
    }

    subQuery := db.Model(&models.OutputTags{}).
        Select("output_id").
        Where("tag_name IN ?", filter.Tags).
        Where("tag_user_id = ?", filter.UserID)

    if filter.TagsQueryMode == defs.QueryModeAll {
        subQuery = subQuery.Group("output_id").
            Having("COUNT(DISTINCT tag_name) = ?", len(filter.Tags))
    }

    return db.Where("id IN (?)", subQuery)
}
```

i would extract this to separate method

Reply:
Your exact suggestion doesn't work - however, I've adjusted it and now it looks like this:
```go
func (o *Outputs) tagFilterScope(tx *gorm.DB, filter entity.ListOutputsFilter) func(db *gorm.DB) *gorm.DB {
	return func(query *gorm.DB) *gorm.DB {
		subQuery := tx.Model(&models.OutputTags{}).
			Select("output_id").
			Where("tag_name IN ?", filter.Tags).
			Where("tag_user_id = ?", filter.UserID)

		if filter.TagsQueryMode == defs.QueryModeAll {
			subQuery = subQuery.Group("output_id").Having("COUNT(DISTINCT tag_name) = ?", len(filter.Tags))
		}

		return query.Where("id IN (?)", subQuery)
	}
}
```

BTW: To keep things consistent, I've made similar scope for `listActions/labels`

Reply:
i was thinking about that too .. :)